### ISSUE TITLE
Augur halts data collection, cannot be restarted

### ISSUE THREAD

**Issue by jberkus** [chaoss/augur#1349]
Please help us help you by filling out the following sections as thoroughly as you can.

**Description:**

After some hours of operation and importing data from GitHub, Augur stops, without having filled in all data.

**How to reproduce:**
1. New Augur install via Dockerfile
2. Start Augur container
3. Add 70 repos with ~~ 2000 PRs to Augur
3. Run for 36 hours
4. Sometime during the 36 hours,  all workers quit.
5. Try restarting augur
6. Workers do not resume collecting data
7. Database is left in incomplete state, with around 40% of data imported

Possibly related to this, I'm seeing a bunch of errors in the log:

2021-07-15 21:04:28,848 [PID: 96745] workers.gitlab_issues_worker.47734 [register_task_failure() in worker_base.py:L2591] [ERROR]: Recorded job error in the history table for: {'job_type': 'MAINTAIN', 'models': ['gitlab_issues'], 'display_name': 'gitlab_issues model for url: https://github.com/konveyor/tackle-control
s.git', 'given': {'git_url': 'https://github.com/konveyor/tackle-controls.git'}, 'focused_task': 1, 'worker_id': 'workers.gitlab_issues_worker.47734'}

... which is very broken, because these are Github repos, not Gitlab repos, which may explain the failure.  Not how to fix it, howeer.

**Expected behavior:**

Augur should have continued collecting data until all data was backfilled.

**Log files**

```
********** Logfile: gitlab_issues_worker_47734_collection.log
[144215] workers.gitlab_issues_worker.47734 [INFO] Found max id for msg_id column in the message table: 1699999

[144215] workers.gitlab_issues_worker.47734 [INFO] Beginning the process of GitLab Issue Collection...
[144215] workers.gitlab_issues_worker.47734 [ERROR] Worker ran into an error for task: {'job_type': 'MAINTAIN', 'models': ['gitlab_issues'], 'display_name': 'gitlab_issues model for url: https://github.com/konveyor/create-release-draft.git', 'given': {'git_url': 'https://github.com/konveyor/create-release-draft.git'$, 'focused_task': 1}
[144215] workers.gitlab_issues_worker.47734 [ERROR] Worker was processing this task for -0.017161846160888672 seconds.
[144215] workers.gitlab_issues_worker.47734 [ERROR] Printing traceback...
[144215] workers.gitlab_issues_worker.47734 [ERROR] 'GitLabIssuesWorker' object has no attribute 'query_gitlab_contributors'
[144215] workers.gitlab_issues_worker.47734 [ERROR] Traceback (most recent call last):
File "/usr/local/lib/python3.8/site-packages/workers/worker_base.py", line 367, in collect
model_method(message, repo_id)
File "/usr/local/lib/python3.8/site-packages/workers/gitlab_issues_worker/gitlab_issues_worker.py", line 47, in gitlab_issues_model
self.query_gitlab_contributors(task, repo_id)
AttributeError: 'GitLabIssuesWorker' object has no attribute 'query_gitlab_contributors'

[144215] workers.gitlab_issues_worker.47734 [INFO] This task inserted 0 tuples before failure.
[144215] workers.gitlab_issues_worker.47734 [INFO] Notifying broker and logging task failure in database...
[144215] workers.gitlab_issues_worker.47734 [ERROR] Recorded job error in the history table for: {'job_type': 'MAINTAIN', 'models': ['gitlab_issues'], 'display_name': 'gitlab_issues model for url: https://github.com/konveyor/create-release-draft.git', 'given': {'git_url': 'https://github.com/konveyor/create-release-$raft.git'}, 'focused_task': 1, 'worker_id': 'workers.gitlab_issues_worker.47734'}
[144215] workers.gitlab_issues_worker.47734 [INFO] Updated job process for model: gitlab_issues

[144215] workers.gitlab_issues_worker.47734 [INFO] Collection process finished
```

**Software versions:**

    Augur: 0.16.2
    OS: Built using provided Dockerfile from this repo

**Comment by jberkus** [comment]
More info:

I restarted Augur by deleting the original container image and rebuilding it.  This got the workers to start up again, and I'm not seeing the gitlab errors.  However, they are still not collecting data ... in fact, workers visit repos with hundreds of PRs that have not been imported, and conclude that there is nothing to do (see log below).

The move2kube repo in that log has 401 closed PRs and 1 open PR.  Augur imported 1 PR from the repo and then stopped.  When it revists move2kube, it still doesn't "see" the 401 historical PRs.

There was a network restart during the initial import period, and that may correlate with collection stopping.  What this suggests is that if Augur is interrupted during backfilling, for any reason, backfilling stops and cannot be resumed.

Log:

```
********** Logfile: repo_info_worker_50770_collection.log
[276577] workers.repo_info_worker.50770 [INFO] Querying parent info to verify if the repo is forked

[276577] workers.repo_info_worker.50770 [INFO] Updated rate limit, you have: 3886 requests remaining.
[276577] workers.repo_info_worker.50770 [INFO] Querying committers count

[276577] workers.repo_info_worker.50770 [INFO] Updated rate limit, you have: 3884 requests remaining.
[276577] workers.repo_info_worker.50770 [INFO] Inserted info for konveyor/move2kube

[276577] workers.repo_info_worker.50770 [INFO] Worker completed this task in -0.8188269138336182 seconds.

[276577] workers.repo_info_worker.50770 [INFO] Recorded job completion for: {'worker_id': 'workers.repo_info_worker.50770', 'job_type': 'MAINTAIN', 'repo_id': 25430, 'job_model': 'repo_info', 'github_url': 'https://github.com/konveyor/move2kube.git'}

[276577] workers.repo_info_worker.50770 [INFO] Updated job process for model: repo_info

[276577] workers.repo_info_worker.50770 [INFO] Telling broker we completed task: {'worker_id': 'workers.repo_info_worker.50770', 'job_type': 'MAINTAIN', 'repo_id': 25430, 'job_model': 'repo_info', 'github_url': 'https://github.com/konveyor/move2kube.git'}

[276577] workers.repo_info_worker.50770 [INFO] This task inserted: 1 tuples and updated 0 tuples.

[276577] workers.repo_info_worker.50770 [INFO] No job found.
[276577] workers.repo_info_worker.50770 [INFO] Collection process finished

********** Logfile: repo_info_worker_50770_collection.err

********** Logfile: insight_worker_49016_collection.log
[276343] workers.insight_worker.49016 [INFO] Record incomplete history tuple: [6300345]

[276343] workers.insight_worker.49016 [INFO] Calling model method insights_model
[276343] workers.insight_worker.49016 [INFO] Getting table values with the following PSQL query:

SELECT * FROM repo_insights_records WHERE repo_id = 25430


[276343] workers.insight_worker.49016 [INFO] Worker completed this task in -1.139005422592163 seconds.

[276343] workers.insight_worker.49016 [INFO] Recorded job completion for: {'worker_id': 'workers.insight_worker.49016', 'job_type': 'MAINTAIN', 'repo_id': 25430, 'job_model': 'insights', 'git_url': 'https://github.com/konveyor/move2kube.git'}

[276343] workers.insight_worker.49016 [INFO] Updated job process for model: insights

[276343] workers.insight_worker.49016 [INFO] Telling broker we completed task: {'worker_id': 'workers.insight_worker.49016', 'job_type': 'MAINTAIN', 'repo_id': 25430, 'job_model': 'insights', 'git_url': 'https://github.com/konveyor/move2kube.git'}

[276343] workers.insight_worker.49016 [INFO] This task inserted: 0 tuples and updated 0 tuples.

[276343] workers.insight_worker.49016 [INFO] No job found.
[276343] workers.insight_worker.49016 [INFO] Collection process finished
```

**Comment by jberkus** [comment]
going to test blowing away the database today and rebuilding it using v0.17.0

**Comment by jberkus** [comment]
@sgoggins In 0.17.0, data collection does not halt because it never starts.

When I add new repos using the CLI, no worker is spawned to collect data for them.

**Comment by jberkus** [comment]
So, here's what worked in my test:

1. Create new container image on 0.17.0
2. Create schema
3. Add additional repositories to the schema using CLI
4. Start augur-backend with workers
5. Workers start up and collect data for all repos

Here's what didn't work:

1. Create new container image on 0.17.0
2. Create schema
3. Start augur-backend with workers
4. Add additional repos to the schema using CLI
5. Workers never collect data from the new repos

I'll stress that I attemped each of the above once only, so it would be worth someone verifying this diagnosis.

**Comment by IsaacMilarky** [comment]
I was able to verify that the order is important to replicate this issue. It seemed to be able to recognize and use existing data in the database but when I followed your steps to replicate the workers seemed to not fetch data. I will look into this more. For now, it seems the best results happen when repos are added before starting the backend. This should be made less clunky in the future.

I also remember the Gitlab support for the docker container being limited in its present state. The default config doesn't include the gitlab specific workers. I will also try to improve logging so it is more clear _why_ workers quit. 

I will look into this further.

**Comment by jberkus** [comment]
Isaac:

Just for clarity, the repos in my test are GitHub repos.

**Comment by jberkus** [comment]
This issue is being subsumed into the larger "workers halt, no logs" issue here: #1413

**Comment by sgoggins** [comment]
@jberkus : Do you think this is *only* a Docker issue, or are you also encountering it on bare metal?

**Comment by sgoggins** [comment]
We are aware of an issue with the Facade worker after new repositories are added. That worker requires a restart. All the other workers we have tested on bare metal do continue to collect newly added repos after the timeout period has passed, which is specified for each model in the `Housekeeper` block of the `augur.config.json`.

**Comment by sgoggins** [comment]
`timeout` is in seconds.

**Comment by jberkus** [comment]
@sgoggins we thought it was a docker issue, but it appears to be the same issue on bare metal.  We have not yet tested 20.2

**Comment by jberkus** [comment]
Our specific issue appears to be resolved by 20.2 on bare metal.  Dockerfiles are still an issue but that's detailed in #1390.

As such, closing this in favor of the more general work in #1413
